### PR TITLE
Assignment2

### PR DESCRIPTION
**Pull request by 0xlbr**

### CODE REVIEW

File: assignment2/LCA.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	/**
+	* This method is used to find the lowest common ancestor(LSA) of two nodes of a tree
+	* @param root This is the root of the tree.
+	* @param nodeA This is the first node of the given tree.
+	* @param nodeB This is the second node of the given tree.
+	* @return Tree<T> This is the node which is the lowest common ancestor of nodeA and nodeB.
+	*/
+	public static <T> Tree<T> findLCA(Tree<T> root, Tree<T> nodeA, Tree<T> nodeB){
+		
+		if(root == null || root == nodeA || root == nodeB) {
```
Comment (AimeeBorda) [inline assignment2/LCA.java:11]: what if nodeA or nodeB are null or not elements of the tree?

Comment (simon-frankau) [reply assignment2/LCA.java:11]: I'd recommend either making it more robust, or at least commenting on the behaviour in such cases.

File: assignment2/PrintAncestors.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+	* @param tree This is the tree.
+	* @param key This is the key that is being searched for.
+	* @return boolean This is true if the key is found in the tree and false otherwise.
+	*/
+	public static <T> boolean ancestors(Tree<T> tree, T key) {
+		if (tree == null) {
+			return false;
+		} else if (tree.key == key) {
+			return true;
+			// search recursively for the key in the left and right subtrees
```
Comment (simon-frankau) [inline assignment2/PrintAncestors.java:13]: This comment is weirdly formatted, because it refers to the next block, not the block it's in. Maybe pull it out of the whole conditional, to explain the overall algorithm, or put it inside the next block ("having recursively searched for the key, we found it and...").

File: assignment2/PrintAncestors.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+		} else if (ancestors(tree.leftChild, key) || ancestors(tree.rightChild, key)) {
+			// if the key is found in either subtrees, print the current's node key
+			System.out.println(tree.key);
```
Comment (simon-frankau) [inline assignment2/PrintAncestors.java:16]: I would be tempted to make this method generate a list, and then print the list in a wrapper. That way, this function can be easily tested with some unit tests.

File: assignment2/TestLCA.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	@Test
+	public void testsSiblingsLCA() {
+		assertEquals(t2, LCA.findLCA(root, t5, t6));
+	}
+
+	@Test
+	public void testDifferentLevelNodesLCA() {
+		assertEquals(root, LCA.findLCA(root, t3, t2));
+	}
+}
```
Comment (simon-frankau) [inline assignment2/TestLCA.java:61]: Nice set of tests. Looks like good coverage to me (ignoring passing a few more nulls in, which always seems a bit tedious).

File: assignment2/TestPrintAncestors.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	
+	@Test
+	public void testNonExistentKey() {
+		assertEquals(PrintAncestors.ancestors(root, 8), false);
+	}
+	
+	@Test
+	public void testKeyAtRoot() {
+		ByteArrayOutputStream expectedOutput = new ByteArrayOutputStream();
+		System.setOut(new PrintStream(expectedOutput));
```
Comment (simon-frankau) [inline assignment2/TestPrintAncestors.java:35]: Nice way of testing output-printing behaviour! I would still tend to prefer separating the algorithmic and I/O parts, but if you need to test I/O, this is good.

File: assignment2/TestPrintAncestors.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+		assertEquals(PrintAncestors.ancestors(root, 0), true);
+		assertEquals(expectedOutput.toString(), "");
+	}
+	
+	@Test
+	public void testNearLeavesLeft() {
+		ByteArrayOutputStream expectedOutput = new ByteArrayOutputStream();
+		System.setOut(new PrintStream(expectedOutput));
+		assertEquals(PrintAncestors.ancestors(root, 3), true);
+		assertEquals(expectedOutput.toString(), "1" + "\n" + "0" + "\n");
```
Comment (simon-frankau) [inline assignment2/TestPrintAncestors.java:45]: Any reason to not just use e.g. "1\n0\n"?

**Review by simon-frankau — APPROVED** [review]
Looks good. Maybe a bit more corner-case handling, but basically fine with good test coverage.
